### PR TITLE
New package mariadb-c-client; enhance mariadb to interoperate.

### PR DESCRIPTION
**Pull request by chissg**
`mariadb-c-client` is a new package using the distinct, LGPL, MySQL-compatible client library from mariadb.com. It provides the virtual package `mariadb-c`, also provided by an enhanced `mariadb` package for interoperability.

### CODE REVIEW

File: var/spack/repos/builtin/packages/mariadb-c-client/package.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+# This is a template package file for Spack.  We've put "FIXME"
```
Comment (davydden) [inline var/spack/repos/builtin/packages/mariadb-c-client/package.py]: please remove

**Comment by gartung** [comment]
@scheibelp

**Comment by scheibelp** [comment]
I should be able to merge this on 6/11. I was going to make a complaint like https://github.com/spack/spack/pull/7728#issuecomment-395921935, but in this case it looks like you in fact do have multiple packages providing sub-components of the full functionality.

This is still slightly different than the typical `provides` case though. Is the main reason to offer the client separately to save space?

**Comment by chissg** [comment]
@scheibelp  A package with code built requiring the MySQL client libraries can be satisfied either by `mysql` itself, should a recipe exist outside `builtin` (or in the future, perhaps, in it); by `mariadb`, or by the new package `mariadb-c-client`, which is built from a distinct code base to `mariadb`. Here at Fermilab, we are enjoined by policy from providing packages containing servers such as `mysql` or `postgresql` without fulfilling onerous security update management requirements. Additionally (this applies to #7728 also) we have experiments unable to use GPL'd code such as `mariadb`, whereas `mariadb-c-client` is released under the LPGL. A dependent package such as (optionally) Root can nominally use any one of the three without any change in configuration options or source code.

File: var/spack/repos/builtin/packages/mariadb-c-client/package.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+##############################################################################
```
Comment (gartung) [inline var/spack/repos/builtin/packages/mariadb-c-client/package.py]: @chissg This header should be changed to the newer one.

**Comment by chissg** [comment]
@scheibelp Re-upping this for merge: is there any reason not to go ahead with this?

**Comment by chissg** [comment]
@scheibelp Virtual package names have been changed per our discussion on the telecon this afternoon. Please note that this PR does not do _anything_ with respect to the `py-mysqldb1` package or the newly-added `mysql` package (see PR #9978), which will be the subject of a subsequent PR.

File: etc/spack/defaults/packages.yaml [hunk 3]
```diff
@@ -30,9 +30,11 @@ packages:
       java: [jdk, openjdk, ibm-java]
       jpeg: [libjpeg-turbo, libjpeg]
       lapack: [openblas]
+      mariadb-client: [mariadb-c-client, mariadb]
       mkl: [intel-mkl]
       mpe: [mpe2]
       mpi: [openmpi, mpich]
+      mysql-client: [mariadb-c-client, mariadb]
```
Comment (scheibelp) [inline etc/spack/defaults/packages.yaml]: @chissg it looks like you added mariadb as a possible provider here but the `mariadb` package does not have a `provides` for `mysql-client` (only for `mariadb-client`).

Comment (chissg) [reply etc/spack/defaults/packages.yaml]: Yes, the ability of the `mariadb` package to provide a `mysql`-compatible client API has not been verified by me. I will remove it immediately.

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**